pdf/model/fonts/ttfparser.go

```diff
@@ -45,35 +77,80 @@ type TtfType struct {
 	Xmin, Ymin, Xmax, Ymax int16
 	CapHeight              int16
 	Widths                 []uint16
-	Chars                  map[uint16]uint16
+
+	// Chars maps rune values (unicode) to the indexes in GlyphNames. i.e GlyphNames[Chars[r]] is
+	// the glyph corresponding to rune r.
+	Chars map[uint16]uint16
+	// GlyphNames is a list of glyphs from the "post" section of the TrueType file.
+	GlyphNames []string
+}
+
+func (ttf *TtfType) String() string {
+	return fmt.Sprintf("FONT_FILE2{%#q Embeddable=%t UnitsPerEm=%d Bold=%t ItalicAngle=%f "+
+		"CapHeight=%d Chars=%d GlyphNames=%d}",
+		ttf.PostScriptName, ttf.Embeddable, ttf.UnitsPerEm, ttf.Bold, ttf.ItalicAngle,
+		ttf.CapHeight, len(ttf.Chars), len(ttf.GlyphNames))
 }
 
+// ttfParser contains some state variables used to parse a TrueType file.
 type ttfParser struct {
 	rec              TtfType
-	f                *os.File
+	f                io.ReadSeeker
 	tables           map[string]uint32
 	numberOfHMetrics uint16
 	numGlyphs        uint16
 }
 
-// TtfParse extracts various metrics from a TrueType font file.
-func TtfParse(fileStr string) (TtfRec TtfType, err error) {
-	var t ttfParser
-	t.f, err = os.Open(fileStr)
+// NewFontFile2FromPdfObject returns a TtfType describing the TrueType font file in PdfObject `obj`.
+func NewFontFile2FromPdfObject(obj core.PdfObject) (rec TtfType, err error) {
+	obj = core.TraceToDirectObject(obj)
+	streamObj, ok := obj.(*core.PdfObjectStream)
+	if !ok {
+		common.Log.Debug("ERROR: FontFile2 must be a stream (%T)", obj)
+		err = core.ErrTypeError
+		return
+	}
+	data, err := core.DecodeStream(streamObj)
+	if err != nil {
+		return
+	}
+
+	// fmt.Println("===============&&&&===============")
+	// fmt.Printf("%#q", string(data))
+	// fmt.Println("===============####===============")
+
+	t := ttfParser{f: bytes.NewReader(data)}
+	rec, err = t.Parse()
+	return
+}
+
+// NewFontFile2FromPdfObject returns a TtfType describing the TrueType font file in disk file `fileStr`.
+func TtfParse(fileStr string) (rec TtfType, err error) {
+	f, err := os.Open(fileStr)
 	if err != nil {
 		return
 	}
+	defer f.Close()
+
+	t := ttfParser{f: f}
+	rec, err = t.Parse()
+	return
+}
+
+// NewFontFile2FromPdfObject returns a TtfType describing the TrueType font file in io.Reader `t`.f.
+func (t *ttfParser) Parse() (TtfRec TtfType, err error) {
+
 	version, err := t.ReadStr(4)
 	if err != nil {
 		return
 	}
 	if version == "OTTO" {
-		err = fmt.Errorf("fonts based on PostScript outlines are not supported")
+		err = errors.New("fonts based on PostScript outlines are not supported")
 		return
 	}
+	// XXX: !@#$ Not sure what to do here. Have seen version="true"
```

Comment:
Add exception for "true" but then test if specific file works. Might be more differences

Reply:
I ran this on my full corpus and couldn't reproduce it. It must have been a bug in the reader.